via/services/youtube_transcript.py

```diff
@@ -0,0 +1,148 @@
+
+
+#: The preferences that we apply when choosing a default transcript for a video.
+DEFAULT_TRANSCRIPT_PREFERENCES = (
+    {
+        "language_code": re.compile("en"),
+        "name": re.compile("English"),
+        "autogenerated": False,
+    },
+    {
+        "language_code": re.compile("en-.*"),
+        "name": re.compile(r"English \(.*\)"),
+        "autogenerated": False,
+    },
+    {
+        "language_code": re.compile("en"),
+        "name": re.compile(".*"),
+        "autogenerated": False,
+    },
+    {
+        "language_code": re.compile("en-.*"),
+        "name": re.compile(".*"),
+        "autogenerated": False,
+    },
+    {
+        "language_code": re.compile("en"),
+        "name": re.compile(".*"),
+        "autogenerated": True,
+    },
+    {
+        "language_code": re.compile("en-"),
+        "name": re.compile(".*"),
+        "autogenerated": True,
+    },
+)
+
+
+class YouTubeTranscriptService:
+    """A service for getting text transcripts of YouTube videos."""
+
+    def __init__(self, http_service: HTTPService):
+        self._http_service = http_service
+
+    def get_transcript_infos(self, video_id: str) -> List[TranscriptInfo]:
+        """Return the list of available transcripts for `video_id`."""
+        response = self._http_service.post(
+            "https://youtubei.googleapis.com/youtubei/v1/player",
+            json={
+                "context": {
+                    "client": {
+                        "hl": "en",
+                        "clientName": "WEB",
+                        "clientVersion": "2.20210721.00.00",
+                    }
+                },
+                "videoId": video_id,
+            },
+        )
+        json = response.json()
+        dicts = json["captions"]["playerCaptionsTracklistRenderer"]["captionTracks"]
+
+        return [
+            TranscriptInfo(
+                language_code=caption_track_dict["languageCode"].lower(),
+                autogenerated=caption_track_dict.get("kind", None) == "asr",
+                name=caption_track_dict["name"]["simpleText"],
+                url=caption_track_dict["baseUrl"],
+            )
+            for caption_track_dict in dicts
+        ]
+
+    def pick_default_transcript(
+        self, transcript_infos: List[TranscriptInfo]
+    ) -> TranscriptInfo:
+        """Return a choice of default transcript from `transcript_infos`."""
+
+        def matches(transcript_info: TranscriptInfo, preference: dict):
+            for key in ("language_code", "name"):
+                if not preference[key].fullmatch(getattr(transcript_info, key)):
+                    return False
+
+            return preference["autogenerated"] == transcript_info.autogenerated
+
+        for preference in DEFAULT_TRANSCRIPT_PREFERENCES:
+            for transcript_info in transcript_infos:
+                if matches(transcript_info, preference):
+                    return transcript_info
+
+        return transcript_infos[0]
```

Comment:
This is our algorithm for picking the default transcript for a video based on the `DEFAULT_TRANSCRIPT_PREFERENCES` above.

Note that if it doesn't find a match then if falls back on just returning the first transcript.